diff --git libsemanage-2.6/utils/semanage_migrate_store libsemanage-2.6/utils/semanage_migrate_store
index 0ebd285..4339372 100755
--- libsemanage-2.6/utils/semanage_migrate_store
+++ libsemanage-2.6/utils/semanage_migrate_store
@@ -16,7 +16,7 @@ try:
 	import selinux
 	import semanage
 except:
-	print("You must install libselinux-python and libsemanage-python before running this tool", file=sys.stderr)
+	print("You must install libselinux-python3 and libsemanage-python3 before running this tool", file=sys.stderr)
 	exit(1)
 
 
